### PR TITLE
Fixing change access codes

### PR DESCRIPTION
**Pull request by quichey**
Made the changes persist, and added flash messages for successful and unsuccessful changes

### CODE REVIEW

File: db/seeds.rb [hunk 1]
```diff
@@ -6,12 +6,6 @@
 #   cities = City.create([{ name: 'Chicago' }, { name: 'Copenhagen' }])
 #   Mayor.create(name: 'Emanuel', city: cities.first)
 
-User.create!(first: 'Michael', last: 'Wu', email: 'michaelwu@berkeley.edu', team: 'exec', skillset: 'None', sid: '42042069', password: '123456', password_confirmation: '123456', code: User.registration_code)
```
Comment (ckvan) [inline db/seeds.rb:9]: I think there should be at least one member account to login to and test functionalities.

File: app/models/code.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  validates :code, uniqueness: true
+  validate :correct_type
+
+  def correct_type
+    if self.code_type != "regular" && self.code_type != "admin"
+      errors.add(:code_type, "-- Not valid access code type")
+    end
+  end
+
+  def self.regular_code
```
Comment (ckvan) [inline app/models/code.rb]: Some redundancy between regular and admin versions of methods, should consider refactoring in future.

File: db/seeds.rb [hunk 3]
```diff
@@ -6,6 +6,9 @@
 #   cities = City.create([{ name: 'Chicago' }, { name: 'Copenhagen' }])
 #   Mayor.create(name: 'Emanuel', city: cities.first)
 
+Code.create!(code_type: "regular", code: ENV["ACCESS_CODE"])
```
Comment (ckvan) [inline db/seeds.rb]: Reasonable separation of access codes from user model.

File: db/seeds.rb [hunk 4]
```diff
@@ -9,10 +9,10 @@
...
-User.create!(first: 'Jason', last: 'Bi', email: 'bi.jason13@berkeley.edu', team: 'kiwi', skillset: 'None', sid: '87654721', password: 'pancakes', password_confirmation: 'pancakes', code: User.admin_code)
-User.create!(first: 'Chau', last: 'Van', email: 'c.van@berkeley.edu', team: 'kiwi', skillset: 'None', sid: '87654921', password: 'pancakes', password_confirmation: 'pancakes', code: User.admin_code)
-User.create!(first: 'Michael', last: 'Wu', email: 'wuxiaohua1011@berkeley.edu', team: 'kiwi', skillset: 'None', sid: '47654321', password: 'pancakes', password_confirmation: 'pancakes', code: User.admin_code)
-User.create!(first: 'Nick', last: 'Cai', email: 'ncaia@berkeley.edu', team: 'kiwi', skillset: 'None', sid: '87654301', password: 'pancakes', password_confirmation: 'pancakes', code: User.admin_code)
+User.create!(first: 'Kyle', last: 'Ngo', email: 'kylengo357@berkeley.edu', team: 'kiwi', skillset: 'None', sid: '87654321', password: 'pancakes', password_confirmation: 'pancakes', code: Code.admin_code)
```
Comment (ckvan) [inline db/seeds.rb:12]: There should still be at least one general user account available to access.

File: app/models/code.rb (hunk 2 above)
Comment (ckvan) [reply app/models/code.rb]: Redundancy has been resolved here.

File: app/models/code.rb [hunk 5]
```diff
@@ -3,12 +3,16 @@ class Code < ActiveRecord::Base
...
   def correct_type
     if self.code_type != "regular" && self.code_type != "admin"
       errors.add(:code_type, "-- Not valid access code type")
     end
   end
 
+
+  #Strings for flash messages
   def self.code_uniqueness_message
     "Code-change failed: Access codes must be different from each other."
```
Comment (ckvan) [inline app/models/code.rb:15]: It's good that messages became more specific.

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -52,7 +52,7 @@ def edit
 
   def registration_code
     begin
-      flash[:notice] = User.change_registration_code(params[:registration_code])
+      flash[:notice] = User.change_code("regular", params[:registration_code])
```
Comment (ckvan) [inline app/controllers/users_controller.rb:55]: Good, thank you for refactoring and removing that redundancy between regular and admin codes.

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -62,22 +62,11 @@ def registration_code
...
-      flash[:notice] = User.change_admin_code(params[:admin_code])
+      flash[:notice] = User.change_code("admin", params[:admin_code])
     rescue ActiveRecord::RecordInvalid => e
       flash[:notice] = Code.code_uniqueness_message
     end
 
     redirect_to users_admin_path
   end
-
-  def login
```
Comment (ckvan) [inline app/controllers/users_controller.rb:63]: I was wondering about this. I suppose Devise handles them, so they're not needed.

File: app/models/code.rb [hunk 8]
```diff
@@ -3,57 +3,46 @@ class Code < ActiveRecord::Base
...
-    if Code.where(code_type: "regular")[0] == nil
-      Code.create!(code_type: "regular", code: ENV["ACCESS_CODE"])
+  ##Getter method -- only valid types are "regular" and "admin" -- for generalizability
+  def self.get_code(type)
+    if Code.where(code_type: type)[0] == nil
+      Code.create!(code_type: type, code: ENV["#{type.upcase}_CODE"])
     end
 
-    return Code.where(code_type: "regular")[0].code
+    return Code.where(code_type: type)[0].code
```
Comment (ckvan) [inline app/models/code.rb:32]: Better specification than a string indeed.

File: app/models/code.rb [hunk 9]
```diff
@@ -3,57 +3,46 @@ class Code < ActiveRecord::Base
...
+  def self.admin_code
+    Code.get_code("admin")
   end
 
-  def self.set_admin_code(newCode)
-    code = Code.where(code_type: "admin")[0]
-    code.code = newCode
-    code.save!
+  ##Setter_method
+  def self.set_code(type, newCode)
```
Comment (ckvan) [inline app/models/code.rb:45]: I appreciate the collapsing of redundant functions, I really do.

File: spec/models/user_spec.rb [hunk 10]
```diff
@@ -13,14 +13,14 @@
   #tests for change-access-code feature
   it "changes the registration code" do
     newCode = Code.regular_code + "nonsense"
-    User.change_registration_code(newCode)
+    User.change_code("regular", newCode)
```
Comment (ckvan) [inline spec/models/user_spec.rb:16]: You started passing a "type" into your changing access code functions, but there is a string still passed in here?

**Review by ckvan — APPROVED** [review]

File: spec/models/user_spec.rb (hunk 10 above)
Comment (quichey) [reply spec/models/user_spec.rb:16]: the type variable is assumed to be string, since the code_type field of the Code model is a t.string

**Pull request merged**